Sources/NIO/EventLoopFuture.swift

```diff
@@ -891,6 +894,42 @@ extension EventLoopFuture {
         
         return body
     }
+    
+    /// Returns a new `EventLoopFuture` that fires only when all the provided futures complete.
+    /// The new `EventLoopFuture` contains the result of combining the `initialResult` with the
+    /// values of the `[EventLoopFuture<U>]`. This funciton is analogous to the standard library's
+    /// reduce(into:), which does not make copies of the result type for each EventLoopFuture.
+    ///
+    /// The returned `EventLoopFuture` will fail as soon as a failure is encountered in any of the
+    /// `futures`. However, the failure will not occur until all preceding
+    /// `EventLoopFutures` have completed. At the point the failure is encountered, all subsequent
+    /// `EventLoopFuture` objects will no longer be waited for. This function therefore fails fast: once
+    /// a failure is encountered, it will immediately fail the overall EventLoopFuture.
+    ///
+    /// - parameters:
+    ///     - initialResult: An initial result to begin the reduction.
+    ///     - futures: An array of `EventLoopFuture` to wait for.
+    ///     - eventLoop: The `EventLoop` on which the new `EventLoopFuture` callbacks will fire.
+    ///     - combiner: The bifunction used to combine partialResults with new elements.
+    /// - returns: A new `EventLoopFuture` with the combined value.
+    public static func reduce<U>(into initialResult: T, _ futures: [EventLoopFuture<U>], eventLoop: EventLoop, combiner: @escaping (inout T, U) -> Void) -> EventLoopFuture<T> {
+        let p0: EventLoopPromise<T> = eventLoop.newPromise()
+        var result: T = initialResult
+        
+        let future = EventLoopFuture<Void>.reduce((), futures, eventLoop: eventLoop) { (_: Void, value: U) in
```

Comment:
It is worth investigating the performance of this implementation. It is technically correct, but might not be as performant as a direct call to `future.fold` (as done in the non-mutaing `reduce`).